Use target for drivers in Linux systemd (updated) and SMF instances for drivers in new Solaris-like systems (and bring AIX initscript into better shape)

jimklimov opened the pull request:
Includes https://github.com/networkupstools/nut/pull/229 by @miska and some subsequent improvements.

jimklimov commented:
@aquette pointed out a user-experience problem with this approach, which can be solved technically:

> the main point why I'm still not ack'ing that PR is that there are additional user actions requires, compared to the history approach of enabling/starting all drivers in one shot: user has to spawn all sub units, which can be done through some helper script that iterates over ups.conf and does the spawning / enabling.

So indeed such an action might be done as an ExecStartPre= handler for example, or even as another unit to depend on (which might be more easily disabled or masked by the end-user, if not desired).

peterhoeg commented:
@jimklimov , you probably want to look into generators at least for systemd instead of your ```ExecStartPre``` hook.

```shell
man 7 systemd.generator
```

Also, the recommended way to deal with the ```network-online.target``` issue is not to make a copy, but instead a drop-in configuration fragment which can be done without changing the unit file you are providing:

So in ```/etc/systemd/system/nut-driver.target.d/network.conf``` you would put:

``` ini
[Unit]
Requires=network-online.target
After=network-online.target
```

jimklimov commented:
@peterhoeg : Thanks for the hints, especially about https://www.freedesktop.org/software/systemd/man/systemd.generator.html

However I am not sure it is a good match for this use-case, due to this:

> Units written by generators are removed when the configuration is reloaded. That means the lifetime of the generated units is closely bound to the reload cycles of systemd itself. 

One application of NUT in an internal project of ours is for embedded systems, and this regular re-generation and removal of files will cause needless wear on the flash storage. Also if `systemctl` may not be called (per doc) at this stage, the generator script write should apparently resort to creating the `/etc/systemd/...wants...` explicitly which is an architectural flaw of systemd I believe - if there's an API and contract, consumers like us should not crawl under the hood (it is not future-proof in particular).

But the script is already mostly there in this PR, I just expect to wrap it into a separate unit that may be enabled or disabled as the end-user would desire :)

peterhoeg commented:
@jimklimov, the generators should not dump data under ```/etc/systemd/system``` but instead ```/run/systemd/system``` which is mounted on ```tmpfs```. This is how mount points are set from ```/etc/fstab``` as an example.

You can of course achieve the same with the ```Pre``` hook although the generators are run much earlier. Besides, you already have the script for the ```Pre``` hook and written code *always* wins over "should probably write it this way". ;-)

jimklimov commented:
@aquette : I revised the newly added script, adding a standard copyright header in particular - can you please revise if it is OK this way?

jimklimov commented:
Rebased over today's master to take advantage of new Travis CI integration.

aquette commented on scripts/systemd/nut-server.service.in:
```diff
@@ -1,10 +1,13 @@
-Wants=nut-driver.service
+Wants=nut-driver.target
+# The `upsd` is a networked service (even if bound to a `localhost`)
+# so it requres that the OS has some notion of networking already.
```
requ**i**res

jimklimov replied:
thanks, fixed

jimklimov commented:
This had to step beyond original system scope to ensure the solution core can reasonably adapt to different service frameworks. Then to ensure I can install it in those other systems :)

jimklimov commented:
Currently found caveats that might be solved later but require considerable effort:
* Solaris SMF constrains the syntax valid strings for instance names (e.g. not starting with a digit, no period chars) which blocks creation of some UPS driver instances. This might be worked around by hashing the device name e.g. to base64 (and un-hashing instance name when calling upsdrvctl), but is not quite user-friendly. Also can store device name in a service attribute while mangling the instance name to a valid subset. Comparisons (if devices are already wrapped) becomes more complicated in the script in either case.
* Dummy-UPS services that "proxy" another locally defined section are essentially a circular dependency for upsd. While the system might start-up lacking a driver, there should be some timer to re-enable failed not-disabled drivers (would be useful in any case though).

jimklimov commented:
Adding to my previous comment, it currently seems reasonable to provide the section-name conversion routines in NUT common code (e.g. config parser) so tools like upsdrvctl can be called to generate the base64 string, as well as all tools can interpret it as identification of a config section, local or remote monitored NUT device, etc.

Also, service instances created and tracked by the enumeration script should include somewhere the copy or hash of configuration variables used when the service instance was created - e.g. if the driver or even port value was changed for the same-named section, we may have new dependencies to track for service management (destroy and recreate the nut-driver unit instance).

aquette commented on configure.ac:
```diff
@@ -1370,6 +1457,7 @@ conftemp="${datadir}"
 eval conftemp=\"${conftemp}\"
 eval conftemp=\"${conftemp}\"
 DATADIR=${conftemp}
+NUT_DATADIR=${conftemp}
```
typo: variable renamed to NUT_DATADIR, but the below (and probably other places) AC_DEFINE_UNQUOTED was not updated

aquette commented on scripts/systemd/nut-driver-enumerator.sh.in:
```diff
@@ -0,0 +1,444 @@
+    # particular system's physics, both serial and network media may need USB).
+    CURR_DRV="`upsconf_getDriver "$1"`" || return $?
+    case "$CURR_DRV" in
+        *netxml*|*snmp*|*ipmi*|*powerman*|*-mib*|*avahi*|*apcsmart*|*apcupsd*)
```
apcsmart is serial only

jimklimov replied:
Thanks, fixed

jimklimov commented:
Rebased over current master branch state today

jimklimov commented:
Updated to handle service instance names not accepted by the current framework "as is" (using NUT device names) to make an MD5 hash of the name and use that as a fallback; tested with Solaris SMF.

Added an `upsdrvsvcctl` script to manage UPSes via their service instances, if already set up (using original device names - even if the service name had to be MD5-mangled).

jimklimov commented:
@clepple @aquette : Please review if this PR is now suitable to get into 2.7.5 ?

jimklimov commented:
PR updated with documentation and some finesse touches on the scripts. Also made use of common service naming for Linux and Solaris cases.

jimklimov commented:
Any chance/plan to merge this into a pending release, as a NUT usability improvement?

clepple commented:
@jimklimov it sounds like a good idea overall, though I am certainly not an expert on either Solaris-style services or systemd.

I haven't had a chance to set up a test environment for this branch, but if it isn't there already, would it be good to have the non-MD5-encoded text as a comment in the service files?

Also, I think it would be good to have a man page for `upsdrvsvcctl`, even if it just points to `upsdrvctl` and explains what it does differently.

jimklimov commented:
So far the real device (or rather NUT section) name it is at least logged in enumeration script console and its service wrapper's log/journal.

There is indeed no such comment baked-in for SMF service instances - just nothing standard for that. I might define a schema extension and save the value into a "proprietary" attribute, it might be helpful for people investigating with some low-level system tools that look up all attributes of a service (e.g. `svcprop`) or with even evolution of these scripts, but generally it won't be easily visible anyway.

The script already has CLI options to correlate instance names vs DEVICE names (unhashing the names it uses for service units/instances). Adding a layman value may be helpful, nut might also become confusing as/if it is not the authoritative source of that name (systemd on the other hand seems to provide no extensibility for its data schema... at best might abuse some `Environment` definition line to piggyback this config part).

So maybe it is better to implement this if there is actual user demand:
- [ ] Extend schema for SMF to store DEVICE name value as an optional service attribute
- [ ] Extend enum script to save the DEVICE name in the SMF instances
- [ ] Extend enum script to save the DEVICE name in systemd instances (e.g. via envvar lines in override snippets)

Data-saver for SMF:
https://github.com/jimklimov/nut/blob/457823bd0eb6ec6ec3539d4a5b9a9b8953d45d93/scripts/systemd/nut-driver-enumerator.sh.in#L304

For Linux systemd it is there already: https://github.com/jimklimov/nut/blob/457823bd0eb6ec6ec3539d4a5b9a9b8953d45d93/scripts/systemd/nut-driver-enumerator.sh.in#L401

- [ ] Manpage for new scripts

clepple commented:
Ah, I was looking in the wrong place for correlating the names. Thanks for the explanation.

jimklimov commented:
I've played with several package (re-)installation scenarios, finding ways to make the procedure robust. Hopefully struck the balance at least for Solaris-like systems (OpenIndiana actually).

aquette commented on Makefile.am:
```diff
@@ -139,21 +139,28 @@ setver:
 	@echo "Error: 'make setver' no longer exists."
 	@echo "Edit configure.in to set version number."
```
it's configure.ac now ;)
